pkg/gpu/nvidia/mig/mig.go

```diff
@@ -0,0 +1,357 @@
+	"github.com/golang/glog"
+	"golang.org/x/sys/unix"
+
+	pluginapi "k8s.io/kubelet/pkg/apis/deviceplugin/v1beta1"
+)
+
+// TODO: short term, accept nvidia-smi path through commandline.
+// Long term, use nvml to handle all MIG operations.
+const nvidiaSmiPath = "/usr/local/nvidia/bin/nvidia-smi"
+
+var gpuInstanceToComputeInstanceProfiles = map[int]int{
+	nvml.GPU_INSTANCE_PROFILE_1_SLICE: nvml.COMPUTE_INSTANCE_PROFILE_1_SLICE,
+	nvml.GPU_INSTANCE_PROFILE_2_SLICE: nvml.COMPUTE_INSTANCE_PROFILE_2_SLICE,
+	nvml.GPU_INSTANCE_PROFILE_3_SLICE: nvml.COMPUTE_INSTANCE_PROFILE_3_SLICE,
+	nvml.GPU_INSTANCE_PROFILE_4_SLICE: nvml.COMPUTE_INSTANCE_PROFILE_4_SLICE,
+	nvml.GPU_INSTANCE_PROFILE_7_SLICE: nvml.COMPUTE_INSTANCE_PROFILE_7_SLICE,
+}
+
+// Max number of GPU instances that can be created for each profile.
+// Source: https://docs.nvidia.com/datacenter/tesla/mig-user-guide/#partitioning
+var gpuInstanceProfileMaxCount = map[int]int{
+	nvml.GPU_INSTANCE_PROFILE_1_SLICE: 7,
+	nvml.GPU_INSTANCE_PROFILE_2_SLICE: 3,
+	nvml.GPU_INSTANCE_PROFILE_3_SLICE: 2,
+	nvml.GPU_INSTANCE_PROFILE_4_SLICE: 1,
+	nvml.GPU_INSTANCE_PROFILE_7_SLICE: 1,
+}
+
+var partitionSizeToProfileMap = map[string]int{
+	"1g.5gb":  nvml.GPU_INSTANCE_PROFILE_1_SLICE,
+	"2g.10gb": nvml.GPU_INSTANCE_PROFILE_2_SLICE,
+	"3g.20gb": nvml.GPU_INSTANCE_PROFILE_3_SLICE,
+	"7g.40gb": nvml.GPU_INSTANCE_PROFILE_7_SLICE,
+}
+
+// DeviceManager performs various management operations on mig devices.
+type DeviceManager struct {
+	devDirectory  string
+	procDirectory string
+	devices       []*nvml.Device // Represents whole GPUs
+	gpuPartitions map[string][]pluginapi.DeviceSpec
+}
+
+// NewDeviceManager creates a new DeviceManager to handle MIG devices on the node.
+func NewDeviceManager(devDirectory, procDirectory string) DeviceManager {
+	return DeviceManager{
+		devDirectory:  devDirectory,
+		procDirectory: procDirectory,
+		devices:       make([]*nvml.Device, 0),
+		gpuPartitions: make(map[string][]pluginapi.DeviceSpec),
+	}
+}
+
+// ListGPUPartitionDevices lists all the GPU partitions as devices that can be advertised as
+// resources available on the node.
+func (d *DeviceManager) ListGPUPartitionDevices() map[string]pluginapi.Device {
+	devices := make(map[string]pluginapi.Device)
+
+	for id := range d.gpuPartitions {
+		devices[id] = pluginapi.Device{
+			ID:     id,
+			Health: pluginapi.Healthy,
+		}
+	}
+
+	return devices
+}
+
+// DeviceSpec returns the device spec that inclues list of devices to allocate for a deviceID.
+func (d *DeviceManager) DeviceSpec(deviceID string) ([]pluginapi.DeviceSpec, error) {
+	deviceSpecs, ok := d.gpuPartitions[deviceID]
+	if !ok {
+		return []pluginapi.DeviceSpec{}, fmt.Errorf("invalid allocation request with non-existing GPU partition: %s", deviceID)
+	}
+
+	return deviceSpecs, nil
+}
+
+// Start method performs the necessary initializations and starts the mig.DeviceManager.
+func (d *DeviceManager) Start() error {
+	if err := nvml.Init(); err != nil {
```

Comment:
is it ok to init 2 nvmls? coz there is another nvml running for monitoring GPUs

Reply:
I think it should be fin. All it does is call `dlopen` on libnvml.so. If there are multiple dlopen() calls, we still get the same handle. Eventually we can move the initialization to a common place, but for now just keeping the features isolated.